### PR TITLE
Chosen combined solution by Group 8

### PR DESCRIPTION
**Pull request by muzaluisa**
First version to be refactored by everyone :)
- Movement class is added
- Yarinka's trick is added
- Complexity comments are added

### CODE REVIEW

File: RearrangeCarGroup8Movements.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        self.from_index = from_index
+        self.to_index = to_index
+
+    def __str__(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        return move_str
+
+    def print_log_file(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        logging.info(move_str)
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: Perhaps we don't really need this function? If we want to log the info, we can perhaps do `logging.info(movement_object.__str__())`?

File: RearrangeCarGroup8Movements.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            self.not_arranged_cars.difference_update([car_to_move]) # O(1)
+
+
+        movement = Movement(car_to_move, car_to_move_position, self.empty_position)
+        if self.print_console:
+            print movement
+        else:
+            movement.print_log_file()
+
+        self._move_car_to_empty_slot(car_to_move_position)
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: If we are to return a list of movement objects instead of printing right away, then perhaps we can remove line 151 - 154 and then just return `Movement(car_to_move, car_to_move_position, self.empty_position)` ^^

File: RearrangeCarGroup8Movements.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        if len(self.not_arranged_cars) > 1000:
+            self._create_log_file()
+            self.print_console = False
+
+        '''Final solution is O(n)
+        '''
+
+        while self.not_arranged_cars:
+            self._rearrange_one_car() # O(1)
+
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: What about moving `self._reset_variables_to_init_state()` to the very beginning of this function, so that we it automatically reset all states before checking not arranged cars? (Then we can remove the first `if` and `elif` checks. Since if the init state is equal to the final state, there will be no cars in `self.not_arranged_cars` and en empty list will be returned.) 

And then since we plan to return a movement object from `self._rearrange_one_car()`, we can let this function return a list of movements, and write another function that does the printing of the list?

File: RearrangeCarGroup8Movements.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        num_cars = len(self.not_arranged_cars)
+        info_message = 'Number of cars to arrange is {0} > 1000, movements are logged'.format(num_cars)
+        logging_message = info_message
+        logging.warning(logging_message)
+        logging.info('Initial car order:')
+        logging.info(self.initial_car_order)
+        logging.info('Final car order:')
+        logging.info(self.final_car_order)
+        logging.info('Moves:')
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: I have never used `logging` before, so I am a bit confused... When I was testing from my computer, I couldn't really find the log file, but I saw all the logging info being printed out to the console...

**Review by luxiya01 — COMMENTED** [review]
A few comments regarding the movement class and the log file XD

File: RearrangeCarGroup8Movements.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        Returns:
+            None
+        '''
+
+        self.initial_car_order = init_car_order
+        self.final_car_order = final_car_order
+        self.print_console = True
+        self._variables_to_init_state()
+
+    def print_moves_log_file(self, moves_list):
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: I think we should move printing functions to a new class to separate handling it from the logic (better separation of concerns! :D)

File: RearrangeCarGroup8Movements.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+        
+        info_message = 'Number moves is {0} > 1000, logging them..'.format(len(moves_list))
+        print info_message
+        logging.info('Initial car order:')
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: I'm wondering about putting this in a one function with newlines, as that's quite a lot of repetitive calls

File: RearrangeCarGroup8Movements.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    def print_moves_console(self, moves_list):
+        
+        '''Prints a list of moves
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+        
+        print 'Initial car order:', self.initial_car_order
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: We have here the same string to print as in the logging, one string to share between them would be useful

File: RearrangeCarGroup8Movements.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        print 'Final car order:', self.final_car_order
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: We could use some neat formatting (see f strings), but that applies only to Python 3 code

File: RearrangeCarGroup8Movements.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+class RearrangeCars(object):
+
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Stores and updates min_num_steps_left that shows min number of moves left
+        in order to avoid comparison current and final order lists that takes O(n)
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: We should give an overall complexity here and describe the number of moves we make

File: RearrangeCarGroup8Movements.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+        '''Rearranges all cars and prints the moves
+        Note 1: In case there are more than 1000 moves to print
+        logging into the file is done
+        Note 2: Possible to run the function many times
+
+        '''
+        
+        if self.initial_car_order == self.final_car_order:
+            assert ValueError, 'Cars are already in the final order, no moves are required'
```
Comment (marijja) [inline RearrangeCarGroup8Movements.py]: Already arranged cars is a desired situation, I don't think we should raise an error (I suppose this is raising, as the syntax feels unfamiliar to me - would be great if anyone explained)

File: testRearrangeCarGroup8.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    def setUp(self):
+
+        '''Basic setting that will be used in several tests
+        '''
+        self.init_order = [1, 2, 0, 3]
+        self.final_order = [3, 1, 2, 0]
+        self.rc = RearrangeCars(self.init_order, self.final_order)
+
+    def test_rearrange_all_cars(self):
```
Comment (marijja) [inline testRearrangeCarGroup8.py]: Maybe rearrange_sample_cars? (Minor change, but you know immediately what's up :D)

File: testRearrangeCarGroup8.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+        final_order_same_empty = [3, 1, 0, 2]
+        rc_same_empty_pos = RearrangeCars(self.init_order, final_order_same_empty)
+        rc_same_empty_pos.rearrange_all_cars()
+        self.assertEqual(rc_same_empty_pos.current_car_order, final_order_same_empty)
+
+    def test_long_parking_lot(self):
+
+        '''Runs 30 test case with different initial and final orders
+        of length N to check that the final arrangement is correct
```
Comment (marijja) [inline testRearrangeCarGroup8.py]: num_spots instead of N or reverse :D

**Review by marijja — COMMENTED** [review]
I think we should add some tests (for example trivial cases with empty list or something like that), maybe basing on other solutions from our group

File: RearrangeCarGroup8Movements.py (hunk 4 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: Should work now, I added logging.basicConfig(filename='RearrangeCarGroup8.log', filemode='w', level=logging.DEBUG)

File: RearrangeCarGroup8Movements.py (hunk 9 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: yes I will write that)

File: RearrangeCarGroup8Movements.py (hunk 6 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]:
```
logging.info('Initial car order:', self.initial_car_order)
logging.info('Final car order:', self.final_car_order)
```
Does this look good? or what you would recommend?

File: RearrangeCarGroup8Movements.py (hunk 10 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: Good point, the assert is not needed here probably, I fix that)

File: RearrangeCarGroup8Movements.py (hunk 9 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: @marijja Could you please add tests, if you can?

File: RearrangeCarGroup8Movements.py (hunk 10 above)
Comment (luxiya01) [reply RearrangeCarGroup8Movements.py]: +1 for removing the assert. I think we can get rid of the checking entirely :)

File: testRearrangeCarGroup8.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        self.rc.rearrange_all_cars()
+        self.assertEqual(self.rc.current_car_order, self.final_order)
+
+    def test_already_proper_order(self):
+
+        '''Tests the case when initial and final orders are the same
+        '''
+
+        rc_same = RearrangeCars(self.final_order, self.final_order)
+        self.assertRaises(rc_same.rearrange_all_cars())
```
Comment (luxiya01) [inline testRearrangeCarGroup8.py]: It's nice to test against already ordered input! But since we are planning to remove the assert from the `rearrange_all_cars()`, we can probably change line 44 to `self.assertEqual(self.rc.current_car_order, self.final_order)` ^^

File: RearrangeCarGroup8Movements.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        moves = []
+        while self.not_arranged_cars:
+            moves.append(self._rearrange_one_car()) # O(1)
+        
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
+        else:
+            self.print_moves_console(moves)
+            
+        self._variables_to_init_state()
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: `self._variables_to_init_state()` (line 205) should probably be placed at the beginning of the function `rearrange_all_cars()`, otherwise I supposed all tests will fail since we will then set `current_car_order` to `initial_car_order` XD

File: RearrangeCarGroup8Movements.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+        '''
+
+        moves = []
+        while self.not_arranged_cars:
+            moves.append(self._rearrange_one_car()) # O(1)
+        
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
+        else:
+            self.print_moves_console(moves)
```
Comment (luxiya01) [inline RearrangeCarGroup8Movements.py]: What about making this function returning the moves, and then make another function that takes a list of movements as input parameter and does the actual printing? This isn't really what the assignment asked for, but will allow the result of rearrange_all_cars() to be integrated into larger modules XD

File: RearrangeCarGroup8Movements.py (hunk 8 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: @marijja Ok, I try to change at least this function to format, that is available in 2.7 then?

File: testRearrangeCarGroup8.py (hunk 11 above)
Comment (muzaluisa) [reply testRearrangeCarGroup8.py]: Sounds good!:)

File: RearrangeCarGroup8Movements.py (hunk 14 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: True, fixed that 👍

File: RearrangeCarGroup8Movements.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+from 0 to N. Cars are with numbers from 1 to N-1, 0 means an empty parking space.
+
+E.G. Initial order = [1, 2, 0, 3]
+       Final order = [3, 1, 2, 0]
+
+Task: to rearrange the cars in a given way
+and print a series of moves from position x to y
+
+For our example, the series of moves should be e.g.:
+    move 1 to 2
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: Maybe change this to the actual string we are outputting, ie:
"Move car # 1 from 0 to 1", for consistency

File: RearrangeCarGroup8Movements.py [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        self.print_console = print_console
+
+    def _prepare_order_message(self, initial_car_order, final_car_order):
+        return ('Initial car order: {init_order}'.format(init_order=initial_car_order) +
+                'Final car order: {final_order}'.format(final_order=final_car_order) +
+                'Moves:')
+
+    def print_moves(self, moves):
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: I think we should keep the comment of while we are doing the log and how it improves the code!
If I'm not mistaken, it should be on @muzaluisa previous version.

File: RearrangeCarGroup8Movements.py [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    def __init__(self, init_order, final_order, print_console):
+        self.order_message = self._prepare_order_message(init_order, final_order)
+        self.print_console = print_console
+
+    def _prepare_order_message(self, initial_car_order, final_car_order):
+        return ('Initial car order: {init_order}'.format(init_order=initial_car_order) +
+                'Final car order: {final_order}'.format(final_order=final_car_order) +
+                'Moves:')
+
+    def print_moves(self, moves):
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: I think that in the code we should also deal with the case that the initial_car_order is already the same of the final_car_order and hence no moves have to be done.

File: testRearrangeCarGroup8.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
+import unittest
+import numpy as np
+from RearrangeCarGroup8Movements import RearrangeCars
+
+class TestRearrangeCars(unittest.TestCase):
```
Comment (blmussati) [inline testRearrangeCarGroup8.py]: I think we should also test the case in which init_order and final_order are of different length - how our function handle that case? I believe it should throw an error of the sort "These are not the same car parks!" or "There is a mismatching in the car parks' sizes, please check them again". Something like that.

File: testRearrangeCarGroup8.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+    '''Unittests for RearrangeCars class and promised fuctionality
+    '''
+
+    def setUp(self):
+
+        '''Basic setting that will be used in several tests
+        '''
+        
+        self.init_order = [1, 2, 0, 3]
```
Comment (blmussati) [inline testRearrangeCarGroup8.py]: Should we name them initial_car_order and final_car_order for consistency through the files?

File: RearrangeCarGroup8Movements.py [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+
+class RearrangeCars(object):
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Print moves in O(n), where n is the number of cars.
+       - More precisely at most n + k moves are taken, where k << n is the amount of movements
+       - where car is placed not to the final place
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: Don't understand what you mean with "the amount of movements where car is placed not to the final place"

File: RearrangeCarGroup8Movements.py [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+
+    def __init__(self, init_car_order, final_car_order):
+
+        '''Initializes class instance with initial, final order and
+        stores the position of empty parking space in the current car order
+        Memory is O(n) to store 3 order lists, 1 car_position_dict and a set of
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: helps to **know**, it's just a minor typo though :)

File: RearrangeCarGroup8Movements.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+
+        self.car_position_dict[car_to_move] = self.car_position_dict[0]
+        self.car_position_dict[0] = car_to_move_position
+
+    def _get_not_arranged_car_index(self):
+
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars))  # O(1)
```
Comment (blmussati) [inline RearrangeCarGroup8Movements.py]: Why do we need to use random? Why we create a tuple?

File: testRearrangeCarGroup8.py [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        '''Test whether the class instance can properly handle
+        running the same function twice
+        '''
+
+        self.rc.rearrange_all_cars()
+        self.assertEqual(self.rc.current_car_order, self.final_order)
+        self.rc.rearrange_all_cars()
+        self.assertEqual(self.rc.current_car_order, self.final_order)
+
+    def test_already_proper_order(self):
```
Comment (blmussati) [inline testRearrangeCarGroup8.py]: Sorry, didn't see this before :)

File: testRearrangeCarGroup8.py [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+        '''Tests the case when the position for the empty spot
+        is the same in the initial and final setting
+        '''
+
+        final_order_same_empty = [3, 1, 0, 2]
+        rc_same_empty_pos = RearrangeCars(self.init_order, final_order_same_empty)
+        rc_same_empty_pos.rearrange_all_cars()
+        self.assertEqual(rc_same_empty_pos.current_car_order, final_order_same_empty)
+
+    def test_long_parking_lot(self):
```
Comment (blmussati) [inline testRearrangeCarGroup8.py]: So this is the testing for when moves > 1000?

File: testRearrangeCarGroup8.py [hunk 26]
```diff
@@ -0,0 +1,94 @@
+import unittest
+import numpy as np
+from RearrangeCarGroup8Movements import RearrangeCars
+
+class TestRearrangeCars(unittest.TestCase):
+
```
Comment (blmussati) [inline testRearrangeCarGroup8.py]: Another test worth doing is if one or both inputs given is/are the empty list!
How should we handle such case? If they are both empty, I would just raise the message "No moves required" or something like that. If just one is empty, we should raise an error, I believe.

**Review by blmussati — COMMENTED** [review]
Hi everyone! I left some comments, mainly regarding the testing that we are missing and two questions on things which weren't clear to me. I fixed a typo in a comment, but it's a minor thing.

Please if someone can give a look!

File: testRearrangeCarGroup8.py (hunk 26 above)
Comment (luxiya01) [reply testRearrangeCarGroup8.py]: I think that's handled by `test_empty_parking_lot` XD

File: RearrangeCarGroup8Movements.py (hunk 16 above)
Comment (luxiya01) [reply RearrangeCarGroup8Movements.py]: Good idea! Will fix XD

File: RearrangeCarGroup8Movements.py (hunk 18 above)
Comment (luxiya01) [reply RearrangeCarGroup8Movements.py]: This seems to be handled by the set `self.not_arranged_cars)`. If the two lists are equal, then it wouldn't enter the while-loop and should just return XD

File: RearrangeCarGroup8Movements.py (hunk 21 above)
Comment (luxiya01) [reply RearrangeCarGroup8Movements.py]: I think the last line shall continue from the second-last line, will fix!

File: testRearrangeCarGroup8.py (hunk 19 above)
Comment (luxiya01) [reply testRearrangeCarGroup8.py]: @blmussati I think this sounds like a very good idea! Do you think putting the check in the `__init__` function will be a good idea? Or should we put it in the method `rearrange_all_cars()`?

Comment (luxiya01) [reply testRearrangeCarGroup8.py]: I temporarily put it in `rearrange_all_cars()`, but feel free to change it! :)

File: RearrangeCarGroup8Movements.py (hunk 9 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: This part is done I suppose)

File: RearrangeCarGroup8Movements.py (hunk 23 above)
Comment (muzaluisa) [reply RearrangeCarGroup8Movements.py]: I googled how to pick up in 0(1) the element from the set, this was the proposed solution. Since tuple are immutable I suppose we add tuple operation :)

File: testRearrangeCarGroup8.py [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+        '''Basic setting that will be used in several tests
+        '''
+        
+        self.initial_car_order = [1, 2, 0, 3]
+        self.final_car_order = [3, 1, 2, 0]
+        self.rc = RearrangeCars(list(self.initial_car_order), list(self.final_car_order))
+
+    def test_empty_parking_lot(self):
+
+        '''Test the case when the parking lot is empty,
```
Comment (blmussati) [inline testRearrangeCarGroup8.py:21]: Just for reference, it's a car park, a parking lot is a slot of the car park. I think this test is for when there are no cars parked, am I right?
Should we change the name to clarify?

File: testRearrangeCarGroup8.py (hunk 26 above)
Comment (blmussati) [reply testRearrangeCarGroup8.py]: Ok! I added a comment there now then, for clarification of terminology :)

File: testRearrangeCarGroup8.py (hunk 19 above)
Comment (blmussati) [reply testRearrangeCarGroup8.py]: I cannot see the update but it might be because my git is still connecting to an old version, I'm not sure :/

Comment (blmussati) [reply testRearrangeCarGroup8.py]: I would putit in __init__ but since the deadline is passed I think rearrange_all_cars() will do :D

File: RearrangeCarGroup8Movements.py (hunk 23 above)
Comment (blmussati) [reply RearrangeCarGroup8Movements.py]: Cool!

File: RearrangeCarGroup8Movements.py (hunk 21 above)
Comment (blmussati) [reply RearrangeCarGroup8Movements.py]: Great!

File: RearrangeCarGroup8Movements.py (hunk 18 above)
Comment (blmussati) [reply RearrangeCarGroup8Movements.py]: Ok thanks @luxiya01  :D

**Review by blmussati — COMMENTED** [review]
Sorry, I didn't spot this yesterday! :( anyway the test work, it was just a matter of correct terminology

File: testRearrangeCarGroup8.py (hunk 27 above)
Comment (carmenmitru) [reply testRearrangeCarGroup8.py:21]: Hello @muzaluisa I've seent at testing that you use  unittest.main(verbosity=2).What is the purpose of this parameter?

File: testRearrangeCarGroup8.py [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+    '''
+
+    def setUp(self):
+
+        '''Basic setting that will be used in several tests
+        '''
+        
+        self.initial_car_order = [1, 2, 0, 3]
+        self.final_car_order = [3, 1, 2, 0]
+        self.rc = RearrangeCars(list(self.initial_car_order), list(self.final_car_order))
```
Comment (yarynapavlyshyn) [inline testRearrangeCarGroup8.py:17]: I do not understand why we need here to convert the list to list again.
May we just write 
self.rc = RearrangeCars(self.initial_car_order, self.final_car_order) ?

File: testRearrangeCarGroup8.py [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+
+    def test_empty_parking_lot(self):
+
+        '''Test the case when the parking lot is empty,
+        should work without raising an error
+        '''
+        
+        rc_empty = RearrangeCars(list(), list())
```
Comment (yarynapavlyshyn) [inline testRearrangeCarGroup8.py]: And here it is better to pass just two "[]" parameters as list creation with "[", "]" is significantly faster than by list() and also it will be clearer what order we have and what are willing to obtain.

**Review by yarynapavlyshyn — COMMENTED** [review]
Beautiful, great modularized, easy to read :)

File: testRearrangeCarGroup8.py (hunk 27 above)
Comment (yarynapavlyshyn) [reply testRearrangeCarGroup8.py:21]: @carmenmitru, Hi Carmen, here is the information I found.
The parameter is a verbosity level. There are 3 different levels:
- 0 (quiet): you just get the total numbers of tests executed and the global result
- 1 (default): you get the same plus a dot for every successful test or a F for every failure
- 2 (verbose): you get the help string of every test and the result

File: testRearrangeCarGroup8.py (hunk 28 above)
Comment (luxiya01) [reply testRearrangeCarGroup8.py:17]: I did it here because in the test for initial_car_order, we checked self.rc.initial_car_order against self.car_order. If we don't list it, maybe these two variables will point to the same object?

Comment (muzaluisa) [reply testRearrangeCarGroup8.py:17]: @luxiya01 

Also, I guess that we need to test 
instead 
`self.assertEqual(rc_length_one.current_car_order, rc_length_one.current_car_order)`
this: 
`self.assertEqual(rc_length_one.current_car_order, [0])`
since we want to check if we have final order we want instead of final order that was saved in the class variable (that could be theorically initialized in a wrong way) :)
So should be smth like this:
 ```
def test_parking_lot_length_one(self):

        '''Test the case when the parking lot is of
        length one and there is no car placed in it
        '''
        
        rc_length_one = RearrangeCars([0], [0])
        rc_length_one.rearrange_all_cars(print_console=True)
        self.assertEqual(rc_length_one.current_car_order, [0])
```

File: testRearrangeCarGroup8.py (hunk 29 above)
Comment (muzaluisa) [reply testRearrangeCarGroup8.py]: I will edit that, thanks :)

File: testRearrangeCarGroup8.py (hunk 28 above)
Comment (luxiya01) [reply testRearrangeCarGroup8.py:17]: @muzaluisa Hmm... Or perhaps we can add a test case that checks whether `rc.final_car_order` is the same as `self.final_car_order` after one call of `rearrance_all_cars`?

Comment (muzaluisa) [reply testRearrangeCarGroup8.py:17]: @luxiya01 Do you afraid that is self.initial_car_order changes, then the value of the object changes? 
This will not happen e.g. I tested as shown below :)

```
def setUp(self):

        '''Basic setting that will be used in several tests
        '''
        
        self.initial_car_order = [1, 2, 0, 3]
        self.final_car_order = [3, 1, 2, 0]
        self.rc = RearrangeCars(self.initial_car_order, self.final_car_order)
        
    def test_change_init_order(self):
       print 'change the reference variable'
       self.initial_car_order = [3, 1, 2, 0]
       self.assertEqual(self.rc.initial_car_order, [1, 2, 0, 3])
```

Comment (luxiya01) [reply testRearrangeCarGroup8.py:17]: Yes, that was what I was afraid of, because I think it happened before... Previously, when the code in `RearrangeCarGroup8`contains the line `self.current_car_order = self.initial_car_order`, instead of `self.current_car_order = list(self.initial_car_order)`, these two variables refer to the same object and `self.intial_car_order` was modified. However, when I checked whether `self.rc.initial_car_order` was the same as `self.initial_car_order`, the test passed like a charm. When I printed them out, however, both had changed and became the same as `self.rc.current_car_order`, since they all pointed to the same object. That was why I listed them... OMG, hope this makes sense XD

Comment (luxiya01) [reply testRearrangeCarGroup8.py:17]: @muzaluisa @yarynapavlyshyn So to sum it all up... I listed them so that if these variables were modified by calling out `rearrange_all_cars`, we will be able to detect them. i.e. without the listing, the `test_initial_order_preserved` wouldn't make much sense, since the variables that we are asserting equal to will be the same variable...

Comment (luxiya01) [reply testRearrangeCarGroup8.py:17]: If, however, we don't intend to check whether `self.initial_car_order` and `self.final_car_order` are modified in the production code, then there is probably no need to list stuffs XD

Comment (muzaluisa) [reply testRearrangeCarGroup8.py:17]: @luxiya01 Now I got your explanation, many thanks. Basically indeed we need the line self.current_car_order = list(self.initial_car_order) in _variables_to_init_state(self). My tests also confirmed that. 

```
def test_initial_order_preserved(self):

        '''Tests whether, after having called rearrange_all_cars() once,
        the field self.initial_car_order is still unchanged

        Note: This test must pass for test_rearrange_sample_cars_twice to
        be meaningful
        '''
        
        print 'test for changing the var'
        
        self.initial_car_order = [3, 1, 2, 0]
        self.rc.rearrange_all_cars(print_console=True)
        print 'changed car order', self.rc.initial_car_order
        self.assertEqual(self.rc.initial_car_order, self.initial_car_order)
```
